### PR TITLE
Add new instruction slots to the VM

### PR DESCRIPTION
**Pull request by grjte**
This PR addresses #884.

- [x] update the constraint definitions in the docs
- [x] update the execution trace generation
- [x] update processor tests
- [x] update AIR constraints
- [x] update AIR tests
- [x] update AIR constraint degrees

### CODE REVIEW

File: docs/src/design/stack/op_constraints.md
Comment (bobbinth) [inline docs/src/design/stack/op_constraints.md:1]: In this file, I believe we also need to:

- Update constraint on line 220.
- Add a similar constraint for the "high-degree operations" section.

File: docs/src/design/stack/op_constraints.md [hunk 1]
```diff
@@ -165,42 +167,42 @@ The degree of this flag is $3$, which is acceptable for a selector for degree $5
...
-| `SPLIT`      | $92$         | `101_1100`      | [Flow control ops](../decoder/main.md) | $6$         |
-| `LOOP`       | $94$         | `101_1110`      | [Flow control ops](../decoder/main.md) | $6$         |
+| `HPERM`      | $80$         | `101_0000`      | [Crypto ops](./crypto_ops.md)          | $5$         |
+| `MPVERIFY`   | $81$         | `101_0001`      | [Crypto ops](./crypto_ops.md)          | $5$         |
+| `PIPE`       | $82$         | `101_0010`      | [I/O ops](./io_ops.md)                 | $5$         |
+| `MSTREAM`    | $83$         | `101_0011`      | [I/O ops](./io_ops.md)                 | $5$         |
+| `SPLIT`      | $84$         | `101_0100`      | [Flow control ops](../decoder/main.md) | $5$         |
+| `LOOP`       | $85$         | `101_0101`      | [Flow control ops](../decoder/main.md) | $5$         |
+| `SPAN`       | $86$         | `101_0110`      | [Flow control ops](../decoder/main.md) | $5$         |
+| `JOIN`       | $87$         | `101_0111`      | [Flow control ops](../decoder/main.md) | $5$         |
```
Comment (bobbinth) [inline docs/src/design/stack/op_constraints.md:189]: Should we add `unused` slots here (similar to how we have it on line 101 above)?

**Review by bobbinth — CHANGES_REQUESTED** [review]
Thank you! Looks good! I left a few comments inline.

Also, would be curious how performance changes before/after this change (i.e., by running Fibonacci example for 2^20 cycles).

File: stdlib/tests/crypto/stark/mod.rs [hunk 2]
```diff
@@ -10,6 +10,7 @@ use test_utils::{
 };
 
 #[test]
+#[ignore]
 fn stark_verifier_e2f4() {
```
[the remark this answers is not in the document]
Comment (grjte) [reply stdlib/tests/crypto/stark/mod.rs:14]: Yes. The width of the trace is hard-coded, so once I added a column it was always going to fail. That's why @Al-Kindi-0 opened #905. You can also see it in the minimal failing changeset I created [here](https://github.com/0xPolygonMiden/miden-vm/commit/99b7d8d75ca8f80c33842324c3f6f4e1b6a5efcc)

File: processor/src/decoder/trace.rs [hunk 3]
```diff
@@ -425,14 +425,18 @@ impl DecoderTrace {
...
         debug_assert_eq!(1, (halt_opcode >> 5) & 1);
-        debug_assert_eq!(own_len, self.op_bit_extra_trace.len());
-        self.op_bit_extra_trace.resize(trace_len, ONE);
-        trace.push(self.op_bit_extra_trace);
+        for (i, mut column) in self.op_bit_extra_trace.into_iter().enumerate() {
+            debug_assert_eq!(own_len, column.len());
+            let value = if i == 1 { ONE } else { ZERO };
+            column.resize(trace_len, value);
+            trace.push(column);
+        }
```
[the remark this answers is not in the document]
Comment (grjte) [reply processor/src/decoder/trace.rs:439]: Sure

File: stdlib/tests/crypto/stark/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Al-Kindi-0) [reply stdlib/tests/crypto/stark/mod.rs:16]: Yes! To save time, I will do the necessary modifications needed to make things pass when I integrate the new instruction I am working on now.

**Review by Al-Kindi-0 — APPROVED** [review]
Looks great, thank you @grjte !
One thing that confused me was that the expressions for `e0` and `e1` are swapped compared to the ones in the issue. I think the issue needs a correction but it is not a big deal.

**Comment by grjte** [comment]
> One thing that confused me was that the expressions for e0 and e1 are swapped compared to the ones in the issue.

@Al-Kindi-0 I don't think they are. In the issue, e0 is the degree reduction column for degree 5 constraints and e1 is the degree reduction column for degree 4 constraints, which is what I've done here. Please double-check in case I've been inconsistent somewhere

File: docs/src/design/stack/op_constraints.md [hunk 4]
```diff
@@ -45,20 +45,22 @@ $$
...
-* Operation flags for operations in the second group (with prefix `10`), are computed using only the first $6$ op bits, and thus, their degree is $6$.
-* Operation flags for operations in the third group (with prefix `11`), are computed using only the first $5$ op bits, and we also use the extra register (which is set to $b_6 \cdot b_5$) to reduce the degree by $1$. Thus, the degree of op flags in this group is $4$.
+* Operation flags for operations in the second group (with prefix `100`), are computed using only the first $6$ op bits, and thus their degree is $6$.
+* Operation flags for operations in the third group (with prefix `101`), are computed using all $7$ op bits. We use the extra register $e_0$ (which is set to $b_6 \cdot (1-b_5) \cdot b_4$) to reduce the degree by $2$. Thus, the degree of op flags in this group is $5$.
+* Operation flags for operations in the fourth group (with prefix `11`), are computed using only the first $5$ op bits. We use the extra register $e_1$ (which is set to $b_6 \cdot b_5$) to reduce the degree by $1$. Thus, the degree of op flags in this group is $4$.
```
Comment (Al-Kindi-0) [inline docs/src/design/stack/op_constraints.md:61]: In #884 , the expressions for `e0` and `e1` are the other way around but I think that is a mistake there. This PR has things right, as far as I can see.

Comment (grjte) [reply docs/src/design/stack/op_constraints.md:61]: ah I see - in the picture it's shown the way I did it, but in the written description it's flipped. You can tell I'm a visual person :)

File: air/src/constraints/stack/op_flags/mod.rs [hunk 5]
```diff
@@ -260,14 +311,19 @@ impl<E: FieldElement> OpFlags<E> {
...
         no_shift_flags[12] =
-            no_shift_flags[9] - degree7_op_flags[29] + degree7_op_flags[28] + degree6_op_flags[8];
+            no_shift_flags[9] - degree7_op_flags[29] + degree7_op_flags[28] + degree5_op_flags[0];
         no_shift_flags[13] = no_shift_flags[12];
         no_shift_flags[14] = no_shift_flags[12];
         no_shift_flags[15] = no_shift_flags[12];
 
         // -------------------------- left shift composite flags computation ----------------------
 
+        // flag of END operation shifting stack to the left. Its effect on stack depends if the
```
[the remark this answers is not in the document]
Comment (grjte) [reply air/src/constraints/stack/op_flags/mod.rs]: Yeah, this was just copy-pasted from before to move it to a more logical spot in the file. I've clarified it & grammar-corrected.

**Comment by grjte** [comment]
> Also, would be curious how performance changes before/after this change (i.e., by running Fibonacci example for 2^20 cycles).

Generated execution trace of 72 columns and 1048576 steps in 380 ms
Generated execution trace of 73 columns and 1048576 steps in 386 ms

Generated execution trace of 72 columns and 4194304 steps in 1543 ms
Generated execution trace of 73 columns and 4194304 steps in 1621 ms

**Comment by bobbinth** [comment]
> Generated execution trace of 72 columns and 1048576 steps in 380 ms
> Generated execution trace of 73 columns and 1048576 steps in 386 ms

This looks right. I'd expect that the impact on proof generation would be great than this.

**Review by bobbinth — APPROVED** [review]
All looks good! Thank you!

I am curious how the prover performance delta looks like for before/after this change.

**Comment by grjte** [comment]
Proof generation is about 5% slower, as you had guessed.

Generated execution trace of 72 columns and 1048576 steps in 380 ms
Program with hash eb717abae795fece4b25a2f30a37908c0221ac6fece594fe8a18d77073d2f536 proved in 65387 ms

Generated execution trace of 73 columns and 1048576 steps in 388 ms
Program with hash eb717abae795fece4b25a2f30a37908c0221ac6fece594fe8a18d77073d2f536 proved in 68009 ms

Generated execution trace of 72 columns and 4194304 steps in 1484 ms
Program with hash 32e404aea4d369a6ec049dc57124e7e786c95e1f63fbe40702cfd16072389acd proved in 466532 ms

Generated execution trace of 73 columns and 4194304 steps in 1608 ms
Program with hash 32e404aea4d369a6ec049dc57124e7e786c95e1f63fbe40702cfd16072389acd proved in 499002 ms

**Pull request merged**